dask/dataframe/core.py

```diff
@@ -4777,7 +4777,7 @@ def set_index(
                     )
 
             # Or be a frame directly
-            elif isinstance(other, DataFrame):
+            elif not TYPE_CHECKING and isinstance(other, DataFrame):
```

Comment:
Mypy was rejecting this line without this check I added.

Reply:
Interesting, it seems that it's complaining that this is unreachable because the function annotation only has `str | Series. So the function annotation is lying a bit. I'm actually not sure why this isn't failing on `main`.

You can see that other places in `set_index()` we have `type: ignore[unreachable]` for the same reason. I'd probably be in favor of doing that instead of this `TYPE_CHECKING` check, since I still want to avoid type checking stuff at runtime.